subsys/bluetooth/host/smp.c

```diff
@@ -1056,7 +1064,10 @@ static void bt_smp_br_disconnected(struct bt_l2cap_chan *chan)
 	BT_DBG("chan %p cid 0x%04x", chan,
 	       CONTAINER_OF(chan, struct bt_l2cap_br_chan, chan)->tx.cid);
 
-	k_delayed_work_cancel(&smp->work);
+	/* Channel disconnected callback is always called from a work handler
```

Comment:
From a work handler that isn't the one in `smp->work`?  If it is `smp->work` this will break things.

Reply:
Can confirm that this only gets called from other work handlers. 

In some cases, especially when combined with callbacks to the application, it'd be really hard to track whether we're cancelling the currently running work handler. What is the consequence of this? Does it just not do a cancel, or can this cause fatal errors in the work queue?

Reply:
Attempts to cancel from the running handler will fail, and the work item will remain active.  You're fine as long as you don't attempt to zero out the work structure (as is done in at least one situation, IIRC).  If you do zero out the structure, or re-initialize it, while it's running "bad things" will happen.

Reply:
I see, thanks!

Reply:
Also note https://docs.zephyrproject.org/latest/reference/kernel/threads/workqueue.html#c.k_work_cancel_delayable_sync: if you try to synchronously cancel (or flush) from the work queue's thread while the work item is associated with that work queue (even if it isn't the one running) you probably won't like whatever happens.